onnx/funconnx/op_run.py

```diff
@@ -0,0 +1,508 @@
+    ) -> Any:
+        """
+        Converts an attribute value into a python value.
+        """
+        if att.type == AttributeProto.GRAPH:
+            from .proto_run import ProtoRun  # type: ignore
+
+            return ProtoRun(
+                att.g,
+                opsets=self.run_params["opsets"],
+                verbose=max(0, self.run_params.get("verbose", 0) - 2),
+            )
+        if att.type in OpRun._attribute_conversion_functions:
+            return OpRun._attribute_conversion_functions[att.type](att)  # type: ignore
+        if ref_att is None:
+            raise AttributeError(
+                f"Unable to convert attribute {att.name!r} type {att.type!r} "
+                f"from node type {self.onnx_node.op_type!r}, "
+                f"domain {self.onnx_node.domain!r}\n{att}."
+            )
+        raise AttributeError(
+            f"Unable to convert default value for {ref_att.name!r} type {att.type!r} "
+            f"from node type {self.onnx_node.op_type!r}, "
+            f"domain {self.onnx_node.domain!r}\n{att}\n{ref_att}."
+        )
+
+    def _load_attributes(self) -> None:
+        "Checks and loads attributes."
+        self.has_linked_attribute = False
+        added_attributes = []
+        for att in self.onnx_node.attribute:
+            name = att.name
+            if att.ref_attr_name:
+                value = RefAttrName(att.ref_attr_name)
+                self.has_linked_attribute = True
+            else:
+                value = self._extract_attribute_value(att)
+            setattr(self, name, value)
+            added_attributes.append(name)
+            if att.type == AttributeProto.GRAPH:
+                setattr(
+                    self,
+                    f"_run_{att.name}",
+                    lambda context, value=value: value.run(None, context or {}),
+                )
+
+        if self._schema and self.onnx_node.op_type not in {"Constant"}:
+            for k, v in self._schema.attributes.items():  # type: ignore
+                if not hasattr(self, k):
+                    if getattr(v, "required", True):
+                        raise RuntimeError(
+                            f"Attribute {k!r} is expected based on ONNX specifications "
+                            f"for node {self.onnx_node.op_type!r}."
+                        )
+                    if hasattr(v, "default_value"):
+                        if v.default_value.type == 0:
+                            # default value is undefined, it depends on the inputs
+                            value = None  # type: ignore
+                        else:
+                            value = self._extract_attribute_value(v.default_value, v)
+                        setattr(self, k, value)
+                        added_attributes.append(k)
+        self.attributes_names_ = set(added_attributes)
+
+    @staticmethod
+    def local_inputs(graph: GraphProto) -> List[str]:
```

Comment:
Minor nit: may be `implicit_inputs` would be a better name? I find `local` confusing since these are *not* local (but belong to outer scope).

Reply:
renamed